server/src/main/java/org/apache/druid/client/MetadataSegmentView.java

```diff
@@ -0,0 +1,202 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class polls the coordinator in background to keep the latest published segments.
+ * Provides {@link #getPublishedSegments()} for others to get segments in metadata store.
+ */
+@ManageLifecycle
+public class MetadataSegmentView
+{
+
+  private static final long POLL_PERIOD_IN_MS = 60000;
+  private static final EmittingLogger log = new EmittingLogger(MetadataSegmentView.class);
+
+  private final DruidLeaderClient coordinatorDruidLeaderClient;
+  private final ObjectMapper jsonMapper;
+  private final BytesAccumulatingResponseHandler responseHandler;
+  private final BrokerSegmentWatcherConfig segmentWatcherConfig;
+
+  private final ConcurrentMap<DataSegment, DateTime> publishedSegments = new ConcurrentHashMap<>(1000);
+  private ScheduledExecutorService scheduledExec;
+
+  @Inject
+  public MetadataSegmentView(
+      final @Coordinator DruidLeaderClient druidLeaderClient,
+      final ObjectMapper jsonMapper,
+      final BytesAccumulatingResponseHandler responseHandler,
+      final BrokerSegmentWatcherConfig segmentWatcherConfig
+  )
+  {
+    this.coordinatorDruidLeaderClient = druidLeaderClient;
+    this.jsonMapper = jsonMapper;
+    this.responseHandler = responseHandler;
+    this.segmentWatcherConfig = segmentWatcherConfig;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    scheduledExec = Execs.scheduledSingleThreaded("MetadataSegmentView-Cache--%d");
+    scheduledExec.schedule(new PollTask(), 0, TimeUnit.MILLISECONDS);
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    scheduledExec.shutdownNow();
+    scheduledExec = null;
+  }
+
+  private void poll()
+  {
+    log.info("polling published segments from coordinator");
+    //get authorized published segments from coordinator
+    final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
+        coordinatorDruidLeaderClient,
+        jsonMapper,
+        responseHandler,
+        segmentWatcherConfig.getWatchedDataSources()
+    );
+
+    final DateTime timestamp = DateTimes.nowUtc();
+    while (metadataSegments.hasNext()) {
+      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      // timestamp is used to filter deleted segments
+      publishedSegments.put(interned, timestamp);
+    }
+    // filter the segments from cache whose timestamp is not equal to latest timestamp stored,
+    // since the presence of a segment with an earlier timestamp indicates that
+    // "that" segment is not returned by coordinator in latest poll, so it's
+    // likely deleted and therefore we remove it from publishedSegments
+    publishedSegments.entrySet().removeIf(e -> e.getValue() != timestamp);
```

Comment:
Would you please also add a comment about `publishedSegments` will be eventually consistent with the segments in the coordinator?

Reply:
sure